### ISSUE TITLE
How are Reference Sets and References related?

### ISSUE THREAD

**Issue by david4096** [ga4gh/schemas#553]
The Reference record does not contain a `referenceSetId` field. Is this an intentional omission? How are references and reference sets made to relate?

**Comment by calbach** [comment]
I guess it isn't actually covered in the [documentation](http://ga4gh-schemas.readthedocs.org/en/latest/api/references.html#references-data-model) from what I can find, but the omission is intentional. That is, one `Reference` may belong to one or more `ReferenceSet`s. Some implementations may choose to only serve one `ReferenceSet` : many `Reference`s, which would still be compatible with these semantics.

Additional background can be found [here](https://github.com/ga4gh/schemas/issues/125).

**Comment by diekhans** [comment]
We need to update the documentation to explain the rationale.

**Comment by david4096** [comment]
Thanks both. Also, references are currently separate from the notion of a dataset. Should references be part of a dataset? @mbaudis has given some considerations that "dataset" is the way that the storage costs and access control will be metered. Should reference sets and/or references get a `datasetId`?

**Comment by mbaudis** [comment]
Do't think so; ```Dataset```s should point to the ```ReferenceSet```(s?), but not vv.

**Comment by richarddurbin** [comment]
I agree that the same Reference should be able to be present in many ReferenceSets, and the same ReferenceSet should be able to
be used in many DataSets.

Richard

> On 4 Feb 2016, at 07:23, Michael Baudis <notifications@github.com> wrote:
> 
> Do't think so; Datasets should point to the ReferenceSet(s?), but not vv.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ga4gh/schemas/issues/553#issuecomment-179686194>.
> 




-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by david4096** [comment]
How should this relationship be expressed in the schema? Other records have a `datasetId` to reflect ownership within a dataset. If datasets will have a number of reference sets associated with them, perhaps an array of `referenceSetIds` should be in the `Dataset` record? Similarly, as a single Reference might be said to be in multiple Reference Sets, should this mapping be done by an array of `referenceIds` in the `ReferenceSet` record?

**Comment by calbach** [comment]
I think there are two orthogonal issues being conflated here:

1. Should  `ReferenceSet`s **belong** to a `Dataset` for purposes of access management or billing. I believe this is David's original questionv.

My 2c - probably. It's a requirement if you intend to implement a server which supports ACLs on top of `Reference`s. The downside is that you may have scenarios where you have a `ReadGroupSet` you can access which refers to a `ReferenceSet` you cannot access.

2. Can the data within a `Dataset` have associations with a heterogeneous mix of `ReferenceSet`s, as defined by the fields `ReadGroupSet.referenceSetId`, `VariantSet.referenceSetId`, etc. To be very concrete, is it legal to have a `Dataset` which contains one `ReadGroupSet` mapped to GRCh38, and another mapped to GRCh37.

I think that this has always been allowed and should continue to be allowed.

**Comment by david4096** [comment]
Thanks @calbach. The first question that I hope can be trivially solved: do the schemas need to change to support access control and metering of references?

The second question, is: How does the many-many relationship of references and reference sets get expressed and implemented?

**Comment by calbach** [comment]
> How does the many-many relationship of references and reference sets get expressed

It is expressed via API method. Namely [`SearchReferences` by `referenceSetId`](https://github.com/ga4gh/schemas/blob/master/src/main/resources/avro/referencemethods.avdl#L97).

> and implemented?

Not sure who's implementation you're asking about. If it's the reference server, I don't know the answer.

**Comment by david4096** [comment]
Thanks again @calbach. I'm trying to narrow down how best to document this if it is left up to the implementor. For example, when searching references using a `referenceSetId`, how does a server know which references are meant to belong to that `referenceSetId` if not by a field in the `Reference` record? When adding a new reference to a server how is one meant to express membership to a Reference set?

**Comment by calbach** [comment]
This goes a bit beyond the scope of the current GA4GH API, since there are no write methods, but I can give a rough sketch.

A server presumably has its own internal storage which is capable of capturing this relationship, in addition to any other data which is not part of the GA4GH spec. Since write is not part of the spec, different servers would have different ingestion mechanisms, but here are a few possible ideas:

- Say a server implementer decides only to support storing one `ReferenceSet` to many `Reference`s. Maybe there is an import script run internally which requires that you provide a set of `Reference`s at once and this becomes a single `ReferenceSet`. Internally, the `Reference`s could be stored with a pointer to this `ReferenceSet`.
- Say a server implementer wants to support storing many:many. One possible API for programmatic ingestion would be something like:

```
CreateReference({reference: {name: "foo", length: 200e6, ...}})
CreateReferenceSet({
    referenceSet: {assemblyId: "GRCh38", ...},
    referenceIds: ["ref1", "ref2"]
})
```

**Comment by macieksmuga** [comment]
Assuming the many-many relationship between `ReferenceSet` and `Reference`, is there a good reason why we couldn't store an array of `ReferenceSetId`s in a `Reference` record, and vice versa? I doubt performance would be an issue, as these lists aren't huge (GRCh38 contains around 500-something contigs or so?) while any given reference wouldn't be part of more than a few reference sets (assuming we're referring only to ones that are stored on the server containing the reference). This way, we'd have a clear relationship embedded in the data themselves, and not need to rely on any external mechanisms to retain this mapping.

**Comment by calbach** [comment]
We originally had a `ReferenceSet.referenceIds` field. This was removed in https://github.com/ga4gh/schemas/pull/378. From experience of having implemented that field, It was very ugly and doesn't really buy the client anything except additional bandwidth cost and confusion around how they are meant to be used. Almost every user assumed they should iterate over them and call `GetReference` N times in serial, which is incredibly inefficient.

The inverse is perhaps not as bad, though I'm still not sure it's buying the client anything. Adding this field would constrain implementations as they would now need to perform bidirectional mappings between `Reference` and `ReferenceSet` (currently, only a lookup of `ReferenceSet` -> `Reference` is required to implement `SearchReference(referensetSetId)`). Another downside is that the associated `ReferenceSet`s are not really a **property** of the `Reference` in a many:many model. It should be possible to have a completely immutable `Reference` which gets added to some arbitrary number of `ReferenceSet`s over time.

I think that given the burden this would cause on the implementer, there should be some kind of justification by use case.

**Comment by diekhans** [comment]
If the relationship from Reference to ReferenceSet is not fixed
and immutable, I don't think it's a good idea to have the data
model have a point back to the ReferenceSet.

While we don't require the storage of objects in the data model,
it's good if design it as if someone could.

**Comment by diekhans** [comment]
From a conceptual point of view, I don't see the problem with
ReferenceSet.referenceIds; that is the containment relationship.
For the human assembly, the list is small.  However, the first
cow assembly had 0.5M contigs, so in general, keeping lists like
are less than idea and the query we have is a much better
approach.

However, the objects pointing to their containers seems to be linked
lists all over again.  Unless there is a one-to-one immutable
relationship from object to container, it's going to cause trouble.

This is not the case with references.  There were 14 releases of
GRCh37 (original + 13 patches).  Each patch release includes unmodified
versions of all the existing sequences, plus new sequences.

If we are going to provide full discovery in the API, having a
query to find all the referencesets containing a given reference
would be good.

I am particularly nervous about containing dataset being
referenced by so many objects give it us still a rather fuzzy
concept.  Given the use case for dataset at Google; it's not
clear that a give piece of data will be in the same dataset
at a different institute.
